runtime/compiler/aarch64/runtime/ARM64RelocationTarget.cpp

```diff
@@ -27,67 +27,93 @@ void arm64CodeSync(unsigned char *codeStart, unsigned int codeSize);
 uint8_t *
 TR_ARM64RelocationTarget::eipBaseForCallOffset(uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
-   return NULL;
+   // reloLocation points at the start of the call offset
+   return reloLocation;
    }
 
 void
 TR_ARM64RelocationTarget::storeCallTarget(uintptr_t callTarget, uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
+   // reloLocation points at the start of the call offset, so just store the uint8_t * at reloLocation
+   storePointer((uint8_t *)callTarget, reloLocation);;
    }
 
 uint32_t
 TR_ARM64RelocationTarget::loadCPIndex(uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
-   return 0;
+   // reloLocation points at the cpAddress in snippet, need to find cpIndex location
+   reloLocation += sizeof(uintptr_t);
+   return (uint32_t)(((uintptr_t )loadPointer(reloLocation)) & 0xFFFFFFFFUL); // Mask out lower 32 bits for index
    }
 
 uintptr_t
 TR_ARM64RelocationTarget::loadThunkCPIndex(uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
-   return 0;
+   // reloLocation points at the cpAddress in snippet, need to find cpIndex location
+   reloLocation += sizeof(uintptr_t);
+   return (uintptr_t )loadPointer(reloLocation);
    }
 
 void
 TR_ARM64RelocationTarget::performThunkRelocation(uint8_t *thunkBase, uintptr_t vmHelper)
    {
-   TR_UNIMPLEMENTED();
+   uintptr_t sendTarget = vmHelper;
+   int32_t *thunkRelocationData = (int32_t *)(thunkBase - sizeof(int32_t));
+
+   storeAddressSequence((uint8_t *)vmHelper, thunkBase + *thunkRelocationData, 1);
+
+   arm64CodeSync(thunkBase, *((int32_t*)(thunkBase - 2*sizeof(int32_t))));
    }
 
 uint8_t *
 TR_ARM64RelocationTarget::loadAddressSequence(uint8_t *reloLocation)
    {
-   TR_UNIMPLEMENTED();
-   return NULL;
+   TR_ASSERT(0, "AOT New Relo Runtime: don't expect loadAddressSequence to be called!\n");
+   uintptr_t value = 0;
+   return (uint8_t *)value;
    }
 
 void
 TR_ARM64RelocationTarget::storeAddressSequence(uint8_t *address, uint8_t *reloLocation, uint32_t seqNumber)
    {
-   TR_UNIMPLEMENTED();
+   uintptr_t value = (uintptr_t)address;
+   uint32_t *movInstruction = (uint32_t *) reloLocation;
+   // immediate offset is at bit 20:5
+   *movInstruction = (*movInstruction & (~((uint32_t)0xFFFF << 5))) | ((value & 0x0000FFFF) << 5);
+   *(movInstruction + 1) = (*(movInstruction + 1) & (~((uint32_t)0xFFFF << 5))) | (((value >> 16) & 0x0000FFFF) << 5);
+   *(movInstruction + 2) = (*(movInstruction + 2) & (~((uint32_t)0xFFFF << 5))) | (((value >> 32) & 0x0000FFFF) << 5);
+   *(movInstruction + 3) = (*(movInstruction + 3) & (~((uint32_t)0xFFFF << 5))) | (((value >> 48) & 0x0000FFFF) << 5);
```

Comment:
@0xdaryl could you verify the correctness from an aarch64 encoding pov?

Reply:
Looks OK.